src/rdf4cpp/rdf/Literal.hpp

```diff
@@ -277,6 +313,255 @@ class Literal : public Node {
+    [[nodiscard]] util::TriBool str_ends_with(std::string_view needle) const noexcept;
+
+    /**
+     * @see https://www.w3.org/TR/xpath-functions/#func-ends-with
+     *
+     * @param needle substring to check
+     * @return whether this' lexical form ends with needle (as xsd:boolean) or the null literal if
+     *      - this is not string-like
+     *      - needle is not string-like
+     *      - the language tags of this and needle do not match
+     */
+    [[nodiscard]] Literal str_ends_with(Literal const &needle, NodeStorage &node_storage = NodeStorage::default_instance()) const noexcept;
+
+    /**
+     * @see https://www.w3.org/TR/xpath-functions/#func-upper-case
+     *
+     * @return the upper case version if this' if this is string-like or the null literal if
+     *      - this is not string-like
+     */
+    [[nodiscard]] Literal uppercase(NodeStorage &node_storage = NodeStorage::default_instance()) const noexcept;
+
+    /**
+     * @see https://www.w3.org/TR/xpath-functions/#func-lower-case
+     *
+     * @return the lower case version if this' if this is string-like or the null literal if
+     *      - this is not string-like
+     */
+    [[nodiscard]] Literal lowercase(NodeStorage &node_storage = NodeStorage::default_instance()) const noexcept;
+
+    /**
+     * @see https://www.w3.org/TR/xpath-functions/#func-concat
+     *
+     * @param other other literal to append to this
+     * @return a string-like type that is the concatenation of the lexical forms of this and other
+     */
+    [[nodiscard]] Literal concat(Literal const &other, NodeStorage &node_storage = NodeStorage::default_instance()) const noexcept;
+
+    /**
+     * akin to std::string_view::substr
+     * @see https://en.cppreference.com/w/cpp/string/basic_string_view/substr
+     * @warning 0-based indexing
+     *
+     * @param start position of the first character
+     * @param len requested length of the substring
+     * @return the characters in this whose position P satisfy: P >= start && P < start + len or the null literal if
+     *      - this is not string-like
+     */
+    [[nodiscard]] Literal substr(size_t start,
+                                 size_t len = std::string_view::npos,
+                                 NodeStorage &node_storage = NodeStorage::default_instance()) const noexcept;
+
+    /**
+     * TODO: should both substr version behave identically?
+     * @see https://www.w3.org/TR/xpath-functions/#func-substring
+     * @warning 1-based indexing
+     *
+     * @param start position of the first character
+     * @param len requested length of the substring
+     * @return the characters in this whose position P satisfy: P >= round(start) && P < round(start) + round(len) or the null literal if
+     *      - this is not string-like
+     *      - start is not xsd:double or derived from it
+     *      - len is not xsd:double or derived from it
+     */
+    [[nodiscard]] Literal substr(Literal const &start,
```

Comment:
What's your opinion on the two `substr` overloads having different behaviour?
- `substr(size_t, size_t, NodeStorage &)` behaves like `std::string_view::substr`
- `substr(Literal const &, Literal const &, NodeStorage &)` behaves as per XPATH spec (you should read it, its weird 😅)

Note: my intention for having two overloads for each function was to have one version that is useful when working in "C++-land" and one version that is useful in "RDF-land" e.g. for query evaluation.

Argument for keeping is this way:
The XPATH spec is weird and the first overload is supposed to be useful in "C++-land" and a `substr(double, double)` is not what anyone would expect from a `substr` function in the context of C++. Especially not the 1-based indexing and handling of negative values.

Argument for making them consistent:
Well they are consistent then, but then we have to choose one behaviour and either move the XPATH weirdness into query evaluation or live with not having a proper C++y interface here.

Reply:
Let's keep both substr versions. I agree the definition is wild. Overhead seems little.